c10/core/MemoryFormat.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <c10/core/Backend.h>
+#include <c10/util/Exception.h>
+
+#include <iostream>
+
+// Memory format is not the property of a Tensor. It is the way to tell an
+// operator how the result should be organized in memory and nothing more. That
+// means memory format should never be used as return value for any tensor state
+// interrogation functions (internally and externally).
+//
+// Possible options are:
+//  Any:
+//    An operator can return Tensor with any memory format. This describes the
+//    current behavior of operators.
+//
+//  Preserve:
+//    If any of the input tensors is in channels_last format, operator output
+//    should be in channels_last format
+//
+//  Contiguous:
+//    Regardless of input tensors format, the output should be contiguous Tensor.
+//
+//  ChannelsLast:
+//    Regardless of input tensors format, the output should be in channels_last format.
+
+
+namespace c10 {
+enum class MemoryFormat : int8_t { Any, Preserve, Contiguous, ChannelsLast };
+
+inline std::ostream& operator<<(
+    std::ostream& stream,
+    at::MemoryFormat memory_format) {
+  switch (memory_format) {
+    case MemoryFormat::Any:
+      return stream << "Any";
+    case MemoryFormat::Preserve:
```

Comment:
nit: in general I think we should be overly conservative about what we expose to python.

I don't think you need Preserve at this point.
I don't really think you need Any -- that seems like it can just be done with an optional, and that fits in with moving things to TensorOptions, which are all optionals anyway.  Maybe there's a case where we would want this for default construction reasons, but I don't see it yet.

"Contiguous" is reusing some of our existing terminology that doesn't have standard definitions.  So it makes sense from a historical perspective, but if you are new to PyTorch, it's not clear to me it makes sense.  Should we call this "C_Contiguous"?

Reply:
Thinking more about `Preserve` -- I'm a bit worried about mixing up specifications and states -- `Preserve` is a valid specification but not a valid state.

We have a similar issue with Devices -- it seems fine at the python level to have a single type (for simplicity of the API), but I find it pretty annoying in the C++ level to constantly have to worry about getting a specification that isn't actually a device (i.e. 'cuda').

So, at the C++ level here we could potentially have MemoryFormat and MemoryFormatSpec.

And actually, do we even need preserve?  Here's a proposal:
The parameter specification of memory format for `empty_like` is:
`MemoryFormat? memory_format=None`.

This could be handled in 1 of 2 ways:
1) we just logically use `None` to mean `Preserve` and everything is a `MemoryFormat?`
2) we complicate the parsing a bit and make the parser for `MemoryFormat?` return a `MemoryFormatSpec`.